### PR TITLE
Improve documentation about underlying types

### LINKED ISSUE #427: matrix/mat64: question: re-use of memory for mat64.Vector pointers seems confusing
In the following code, it seems that `vec1` is not re-initialized at each loop. Is that correct, and if so, is that the intended behavior? I was incredibly confused at this behavior for several hours last night while trying to debug a piece of code.

The following tests were ran with `go version go1.7.1 linux/amd64`.

## Re-use of memory example
In the following, since `vec1` is created inside the loop, I would expect each new iteration to start from scratch and overwrite the `vec1` pointer entirely. However, it behaves as if `vec1` was defined once before the start of the loop.
### Code
```go
func main() {
	values := []float64{1, 2, 3}
	for i := 0; i < 3; i++ {
		vec1 := mat64.NewVector(3, values)
		vec2 := mat64.NewVector(3, []float64{4, 5, 6})
		vec1.SubVec(vec1, vec2)
		fmt.Printf("%+v\n", vec1)
	}
}
```
### Output
```
&{mat:{Inc:1 Data:[-3 -3 -3]} n:3}
&{mat:{Inc:1 Data:[-7 -8 -9]} n:3}
&{mat:{Inc:1 Data:[-11 -13 -15]} n:3
```

## Fix for previous situation
```go
func main() {
	for i := 0; i < 3; i++ {
		values := []float64{1, 2, 3}
		vec1 := mat64.NewVector(3, values)
		vec2 := mat64.NewVector(3, []float64{4, 5, 6})
		vec1.SubVec(vec1, vec2)
		fmt.Printf("%+v\n", vec1)
	}
}
```

### Output
```
&{mat:{Inc:1 Data:[-3 -3 -3]} n:3}
&{mat:{Inc:1 Data:[-3 -3 -3]} n:3}
&{mat:{Inc:1 Data:[-3 -3 -3]} n:3}
```

### PR DESCRIPTION
**Pull request by btracey**
Fixes #427

### CODE REVIEW

**Comment by btracey** [comment]
PTAL @christopherrabotin @kortschak

**Comment by ChristopherRabotin** [comment]
LGTM, thanks!

**Comment by btracey** [comment]
Great. I'm going to wait on @kortschak  for approval also.

File: mat64/dense.go [hunk 1]
```diff
@@ -36,24 +36,26 @@ type Dense struct {
 	capRows, capCols int
 }
 
-// NewDense creates a new matrix of type Dense with dimensions r and c.
-// If the mat argument is nil, a new data slice is allocated.
+// NewDense creates a new Dense matrix with r rows and c columns. If
+// len(data) == r*c, data is used as the underlying slice (changes to the matrix
```
Comment (kortschak) [inline mat64/dense.go]: "data is used as the backing slice. ... Changes to elements of the returned Dense will be reflected in data."

Panic warning in the elipsis.

File: mat64/symmetric.go [hunk 2]
```diff
@@ -51,25 +51,29 @@ type MutableSymmetric interface {
 	SetSym(i, j int, v float64)
 }
 
-// NewSymDense constructs an n x n symmetric matrix. If len(mat) == n * n,
-// mat will be used to hold the underlying data, or if mat == nil, new data will be allocated.
-// The underlying data representation is the same as a Dense matrix, except
-// the values of the entries in the lower triangular portion are completely ignored.
-func NewSymDense(n int, mat []float64) *SymDense {
+// NewSymDense creates a new Symmetric matrix with n rows and columns. If
+// len(data) == n*n, data is used as the underlying slice (changes to the matrix
```
Comment (kortschak) [inline mat64/symmetric.go]: "data is used as the backing slice. ... Changes to elements of the returned SymDense will be reflected in data."

File: mat64/triangular.go [hunk 3]
```diff
@@ -91,22 +91,23 @@ func (t TransposeTri) UntransposeTri() Triangular {
...
-// NewTriDense constructs an n x n triangular matrix with the orientation
-// specified by kind.
-// If len(mat) == n * n, mat will be used to hold the underlying data, if
-// mat == nil, new data will be allocated, and will panic if neither of these
-// cases is true.
-// The underlying data representation is the same as that of a Dense matrix,
-// except the values of the entries in the opposite half are completely ignored.
-func NewTriDense(n int, kind matrix.TriKind, mat []float64) *TriDense {
+// NewTriDense creates a new Triangular matrix with n rows and columns. If
+// len(data) == n*n, data is used as the underlying slice (changes to the matrix
```
Comment (kortschak) [inline mat64/triangular.go]: "data is used as the backing slice. ... Changes to elements of the returned TriiDense will be reflected in data."

File: mat64/vector.go [hunk 4]
```diff
@@ -29,8 +29,9 @@ type Vector struct {
 }
 
 // NewVector creates a new Vector of length n. If len(data) == n, data is used
-// as the backing data slice. If data == nil, a new slice is allocated. If
-// neither of these is true, NewVector will panic.
+// as the underlying slice (changes to the vector will affect the input slice).
```
Comment (kortschak) [inline mat64/vector.go]: "data is used as the backing slice. ... Changes to elements of the returned Vector will be reflected in data."

File: mat64/dense.go (hunk 1 above)
Comment (btracey) [inline mat64/dense.go]: Changed the wording, though not exactly what you said.

File: mat64/symmetric.go (hunk 2 above)
Comment (btracey) [inline mat64/symmetric.go]: Changed the wording, though not exactly what you said.

File: mat64/triangular.go (hunk 3 above)
Comment (btracey) [inline mat64/triangular.go]: Changed the wording, though not exactly what you said.

File: mat64/vector.go (hunk 4 above)
Comment (btracey) [inline mat64/vector.go]: Changed the wording, though not exactly what you said.

**Pull request merged**